### PR TITLE
Feat: custom leading builder

### LINKED ISSUE #1611: Ability to add custom icons for lists
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/singerdmx/flutter-quill/issues)

### Use case

It helps to expand the use of lists. You often need to move away from the usual patterns and use something of your own, design lists with the icon you like. 
Besides, this is now a very frequent request from users in different applications.

### Proposal

example

### LINKED ISSUE #1974: Numbered list wrong numeration after increase + decrease indent
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/singerdmx/flutter-quill/issues)

### Flutter Quill version

master

### Steps to reproduce

1. Increase the numbered list indent 3 times, until the numbering style is the same again.
2. For the last list item decrease its indent 3 times.

### Expected results

The last item numeration start from 1.

### Actual results

The last item numeration start from 2.

### Code sample

<details><summary>Code sample</summary>

```dart
[Paste your code here]
```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

Numbered list numeration bug

</details>

### Logs

<details><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### LINKED ISSUE #2131: Make `QuillEditor*Point` more customizable
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/singerdmx/flutter-quill/issues)

### Use case

Looking for a way to achieve the following design:

Screenshot 2024-08-21 at 11 15 36

This is the component I would like to be able to customize:

Screenshot 2024-08-20 at 11 26 18

### What's not possible currently?

- The `1.` should be aligned to `topLeft` inside `QuillEditorNumberPoint` (currently it's topRight)
- The `1.` should have a specific different color than the rest of the text
- The `width` of `QuillEditorNumberPoint` and left padding should be customizable
- The `1.` should be base aligned with the rest of the text ( this looks like a separate bug 🤔 )

### What is happening currently?

`text_bloc.dart` has `_buildLeading` which decides which bullet point widget will be used:
- QuillEditorNumberPoint
- QuillEditorBulletPoint
- QuillEditorCheckboxPoint

This is getting put under `EditableTextLine` (which is a custom `RenderObjectWidget`).

By default this is getting put in a `Container` of predetermined `width` that will align everything to `topEnd`. It looks like baseline alignment is lost, and background color of the container will be enforced.

There is an option to pass a `customWidget` by `QuillEditorConfigurations` > `QuillEditorElementOptions` > `OrderedList`. There is no way however to dynamically build widget based on the index of the bullet point. Moreover the `customWidget` will be just put inside the `Container` prohibiting any changes to the `Container` itself thus limiting bullet list customization.

### What could happen?

There could be an API that allows you to just `buildLeading` yourself. This could be put under respective configuration objects under `QuillEditorElementOptions`. That said, I could imagine having an option to pass:

`leadingBuilder` to `QuillEditorUnOrderedListElementOptions`. `leadingBuilder` would be a function returning a `Widget` that takes all these arguments:

```dart
QuillEditorNumberPoint(
        index: index,
        indentLevelCounts: indentLevelCounts,
        count: count,
        style: defaultStyles.code!.style
            .copyWith(color: defaultStyles.code!.style.color!.withOpacity(0.4)),
        width: _numberPointWidth(fontSize, count),
        attrs: attrs,
        padding: fontSize,
        withDot: false,
      )
```

Futhermore all the helper methods like `double _numberPointWidth(double fontSize, int count)` should either be public static (so that they can be reused), or there should be a way to override them at the configuration level.

### LINKED ISSUE #2138: unordered list style
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/singerdmx/flutter-quill/issues)

### Use case

I hope Quill can support unordered list formats such as the following image.Because this is the format supported by most rich texts, and we only support black dots at present.The interface provided by Quill does not meet our needs to customize it.

### Proposal

image
image

### LINKED ISSUE #2149: Background Color for Bullet Points
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/singerdmx/flutter-quill/issues)

### The question

Not able to spot the backgroundColor property in the QuillEditorOrderedListElementOptions or QuillEditorUnOrderedListElementOptions, 
in version 10.4.5

image

but it has been implement as stated in 
https://github.com/singerdmx/flutter-quill/issues/1562

### PR DESCRIPTION
**Pull request by CatHood0**
## Description

A new property has been added that will allow us to configure the leading of the elements in a list without any limitations.

### Example of a custom bullet list:

This code: 

```dart
 customLeadingBlockBuilder: (node, config) {
     if (config.attribute == Attribute.ul) {
            final indent = (config.attrs[Attribute.indent.key]?.value as int?) ?? 0;
             return Container(
                  alignment: AlignmentDirectional.topEnd,
                  padding: EdgeInsetsDirectional.only(end: config.padding ?? 0),
                  child: Text(
                       indent == 0
                           ? '😃'
                           : indent == 1
                                ? '👽'
                                : indent == 2
                                     ? '♥︎'
                                     : indent == 3
                                         ? '🫸'
                                         : indent == 4
                                             ? '🌏'
                                             : '🤕',
                            style: TextStyle(fontSize: 13),
               ),
          );
     }
     return null;
},
```

Will have this effect on the view:

Captura de pantalla_2024-08-26_12-36-49

## Related Issues

- *Fix #2138*
- *Fix #2149*
- *Fix #1611*
- *Fix  #2131*

## Not related issues

- *Fix #1974*

These issues are not related but was fixed while creating the feature.

- [x] ✨ **New feature:** Adds new functionality without breaking existing features.
- [x] 🛠️ **Bug fix:** Resolves an issue without altering current behavior.
- [ ] 🧹 **Code refactor:** Code restructuring that does not affect behavior.
- [ ] ❌ **Breaking change:** Alters existing functionality and requires updates.
- [ ] 🧪 **Tests:** Adds new tests or modifies existing tests.
- [ ] 📝 **Documentation:** Updates or additions to documentation.
- [ ] 🗑️ **Chore:** Routine tasks, or maintenance.
- [ ] ✅ **Build configuration change:** Changes to build or deploy processes.

## Suggestions

We should stop creating such hardcoded code as it is really getting complicated trying to create default blocks to give users the opportunity to create their own independent components.

### CODE REVIEW

**Comment by singerdmx** [comment]
Is the node inline or block level?

**Comment by singerdmx** [comment]
OK, I got the idea.
Looking forward to see how it turns out.

**Comment by CatHood0** [comment]
> Is the node inline or block level? 

I trying to create to both types. I pass the `TextLine` as an argument of `TextLineNodeBuilder` to the let the user wrap the common TextLine used by the package, but, it cannot be wrapped since `EditableTextLine` expect another type of widget.

**Comment by CatHood0** [comment]
@singerdmx Do you have any idea on how to use `TextLine` wrapped in another type of widget? I saw that `RenderContentProxyBox` has methods that are used specifically for `RenderEditableTextLine`. If not, I'll have to just limit the Feature to being a Custom builder for `Leading` Lists and Code Blocks.

File: lib/src/editor/raw_editor/raw_editor_state.dart [hunk 1]
```diff
@@ -975,7 +976,7 @@ class QuillRawEditorState extends EditorState
     for (final node in doc.root.children) {
       final attrs = node.style.attributes;
 
-      if (prevNodeOl && attrs[Attribute.list.key] != Attribute.ol) {
+      if (prevNodeOl && attrs[Attribute.list.key] != Attribute.ol || attrs.isEmpty) {
```
Comment (CatHood0) [inline lib/src/editor/raw_editor/raw_editor_state.dart:979]: This new condition fix a wrong numeration of the ordered list.

**Comment by CatHood0** [comment]
@EchoEllet how can i get the analizer ignore the deprecated controller param in editor_onfigurations file?

**Comment by EchoEllet** [comment]
> @EchoEllet how can i get the analizer ignore the deprecated controller param in editor_onfigurations file?

Add a comment before the statement:

```
// ignore: deprecated_member_use_from_same_package
```

If this is not what you meant, then I might need more details as I'm in the middle of something else and haven't been able to take a look at this PR for now.

**Comment by CatHood0** [comment]
@EchoEllet thank you. There's no problem, when you have time just take a look.

**Comment by CatHood0** [comment]
I think that i got something😅

https://github.com/user-attachments/assets/5ba371de-148b-4de6-b102-43d9837069c6

**Comment by CatHood0** [comment]
I'll give up for now because the code is very dependent on certain types that don't allow me to quickly figure out how to correctly implement NodeBuilders. I'll try to implement them later, since I don't have much time for now and it's better to leave this PR as a feature so that the list leaders can be modified without any complications.

**Comment by singerdmx** [comment]
Yeah we can leave it and maybe someone else may be able to help

**Comment by singerdmx** [comment]
This is a big change and it is not so easy

File: lib/src/editor/raw_editor/builders/leading_block_builder.dart [hunk 2]
```diff
@@ -0,0 +1,144 @@
+import 'package:flutter/material.dart';
+import '../../../document/attribute.dart';
+import '../../../document/nodes/node.dart';
+import '../../style_widgets/checkbox_point.dart';
+
+typedef LeadingBlockNodeBuilder = Widget? Function(Node, LeadingConfigurations);
+
+class LeadingConfigurations {
```
Comment (vishna) [inline lib/src/editor/raw_editor/builders/leading_block_builder.dart:8]: this class is not exported, probably should be 👀

Comment (CatHood0) [reply lib/src/editor/raw_editor/builders/leading_block_builder.dart:8]: I forgot it. I'll add it

**Comment by vishna** [comment]
@CatHood0 tried this PR to achieve the bullet list I mentioned in other ticket, works much nicer already 🎉 good work 🙇 

Screenshot 2024-08-28 at 15 39 07

My remarks:

- the base width of the leading container is not customizable but calculated based on `EditableTextBlock._numberPointWidth` or just font size
- `LeadingConfigurations.width` - relying on this to make the positioning look acceptable - it's a nullable, can it ever be null though? 🤔 

```dart
      width: isOrdered || isCodeBlock
          ? _numberPointWidth(fontSize, count)
          : isUnordered
              ? fontSize * 2
              : null, // can this happen ?
```

It would be nice to have a custom builder for the width (and then for the deeper levels' indentation width - I didn't figure out yet which bit is adding this )

File: lib/src/editor/raw_editor/builders/leading_block_builder.dart (hunk 2 above)
Comment (vishna) [reply lib/src/editor/raw_editor/builders/leading_block_builder.dart:8]: also I saw other `*Configurations` classes extend `Equatable` - maybe it makes sense to add this here too 🤔

**Comment by CatHood0** [comment]
> It would be nice to have a custom builder for the width (and then for the deeper levels' indentation width - I didn't figure out yet which bit is adding this )

Can't you just use the width that you want? 

And the width of the Leading just affects to the leading and not the entire block.

**Comment by vishna** [comment]
> Can't you just use the width that you want?
> 
> And the width of the Leading just affects to the leading and not the entire block.

That's the point, the bounding box for whatever container I  build with `customLeadingBlockBuilder` seems to be  __predetermined__ and will be `LeadingConfigurations.width` (or more for upper level). I'll try to hack something tomorrow as I want to be able to customize the space available for the `customLeadingBlockBuilder`

**Comment by CatHood0** [comment]
> That's the point, the bounding box for whatever container I build with customLeadingBlockBuilder seems to be predetermined and will be LeadingConfigurations.width (or more for upper level). I'll try to hack something tomorrow as I want to be able to customize the space available for the customLeadingBlockBuilder

as I say, this is used by default to let the default implementation or the leading blocks. You can always pass to you custom widget the width that you want using the `indentLevelCount` to see what should be the width that need to be used.

**Comment by vishna** [comment]
👋  @CatHood0 

made a PR to your PR here: https://github.com/CatHood0/flutter-quill/pull/1 I hope this explains my points better 😅 Also it seems to work just fine, other test pass and I'm able to have this aspect of the library now fully customized 🎉

**Comment by singerdmx** [comment]
@CatHood0 is this ready for merging?

**Comment by vishna** [comment]
@singerdmx @CatHood0 I'd be grateful to have [my changes](https://github.com/CatHood0/flutter-quill/pull/1) also in 🙇

**Comment by CatHood0** [comment]
@singerdmx let me watch the PR maded by @vishna

**Comment by CatHood0** [comment]
@singerdmx this can be merged now-

**Comment by singerdmx** [comment]
@CatHood0 it would be nice if there are some updates to README about this

File: lib/src/editor/style_widgets/number_point.dart [hunk 3]
```diff
@@ -18,7 +16,7 @@ class QuillEditorNumberPoint extends StatelessWidget {
     this.backgroundColor,
   });
 
-  final int index;
+  final String index;
```
[the remark this answers is not in the document]
Comment (CatHood0) [reply lib/src/editor/style_widgets/number_point.dart:19]: I guess it is not exported. I'll need to check it later

File: lib/src/editor/widgets/text/text_block.dart [hunk 4]
```diff
@@ -265,65 +258,86 @@ class EditableTextBlock extends StatelessWidget {
...
+                              .useTextColorForDot ==
+                          true
+                      ? fontColor
+                      : null,
+                )
+              : isCheck
+                  ? null
+                  : defaultStyles.code!.style.copyWith(
+                      color: defaultStyles.code!.style.color!.withOpacity(0.4),
+                    ),
```
[the remark this answers is not in the document]
Comment (CatHood0) [reply lib/src/editor/widgets/text/text_block.dart:295]: Yeah, we will need to put this on another part.

**Pull request merged**